File: app/src/androidTest/java/com/imagegallery/ImageSearchTest.java
package com.imagegallery;

import android.content.Intent;
import android.support.test.rule.ActivityTestRule;
import android.support.test.runner.AndroidJUnit4;

import com.imagegallery.list.ImageListActivity;

import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.action.ViewActions.click;
import static android.support.test.espresso.action.ViewActions.typeText;
import static android.support.test.espresso.contrib.RecyclerViewActions.actionOnItemAtPosition;
import static android.support.test.espresso.matcher.ViewMatchers.withId;

@RunWith(AndroidJUnit4.class)
public class ImageSearchTest {

    @Rule
    public ActivityTestRule<ImageListActivity> activityTestRule = new ActivityTestRule<>(ImageListActivity.class, false, false);

    @Test
    public void shouldDisplayResultsWhenSearchPerformed() {
        activityTestRule.launchActivity(new Intent());

        // when I type a search term
        onView(withId(R.id.search_view))
                .perform(typeText("dogs"));

        // check there is a single result that can be clicked
        onView(withId(R.id.images_list))
                .perform(actionOnItemAtPosition(0, click()));
    }

}

File: app/src/androidTest/java/com/imagegallery/ImageSearchTest.kt
package com.imagegallery

import android.content.Intent
import android.support.test.espresso.Espresso.onView
import android.support.test.espresso.action.ViewActions.click
import android.support.test.espresso.action.ViewActions.typeText
import android.support.test.espresso.contrib.RecyclerViewActions.actionOnItemAtPosition
import android.support.test.espresso.matcher.ViewMatchers.withId
import android.support.test.rule.ActivityTestRule
import android.support.test.runner.AndroidJUnit4
import android.support.v7.widget.RecyclerView
import com.imagegallery.list.ImageListActivity
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class ImageSearchTest {

    @Rule
    var activityTestRule = ActivityTestRule(ImageListActivity::class.java, false, false)

    @Test
    fun shouldDisplayResultsWhenSearchPerformed() {
        activityTestRule.launchActivity(Intent())

        // when I type a search term
        onView(withId(R.id.searchView))
                .perform(typeText("dogs"))

        // check there is a single result that can be clicked
        onView(withId(R.id.imagesList))
                .perform(actionOnItemAtPosition<RecyclerView.ViewHolder>(0, click()))
    }

}